packages/pubsub/src/subscription.js

```diff
@@ -403,90 +247,121 @@ Subscription.formatName_ = function(projectId, name) {
 };
 
 /**
- * Generate a random name to use for a name-less subscription.
+ * Acks the provided message. If the connection pool is absent, it will be
+ * placed in an internal queue and sent out after 1 second or if the pool is
+ * re-opened before the timeout hits.
  *
  * @private
+ *
+ * @param {object} message - The message object.
  */
-Subscription.generateName_ = function() {
-  return 'autogenerated-' + uuid.v4();
+Subscription.prototype.ack_ = function(message) {
+  this.breakLease_(message);
+  this.histogram.add(Date.now() - message.received);
+
+  if (!this.connectionPool) {
+    this.inventory_.ack.push(message.ackId);
+    this.setFlushTimeout_();
+    return;
+  }
+
+  var self = this;
+
+  this.connectionPool.acquire(message.connectionId, function(err, connection) {
+    if (err) {
+      self.emit('error', err);
+      return;
+    }
+
+    connection.write({ ackIds: [message.ackId] });
+  });
 };
 
 /**
- * Acknowledge to the backend that the message was retrieved. You must provide
- * either a single ackId or an array of ackIds.
+ * Breaks the lease on a message. Essentially this means we no longer treat the
+ * message as being un-acked and count it towards the flow control limits.
  *
- * @resource [Subscriptions: acknowledge API Documentation]{@link https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions/acknowledge}
+ * If the pool was previously paused and we freed up space, we'll continue to
+ * recieve messages.
  *
- * @throws {Error} If at least one ackId is not provided.
+ * @private
+ *
+ * @param {object} message - The message object.
+ */
+Subscription.prototype.breakLease_ = function(message) {
+  var messageIndex = this.inventory_.lease.indexOf(message.ackId);
+
+  this.inventory_.lease.splice(messageIndex, 1);
+  this.inventory_.bytes -= message.data.length;
```

Comment:
One subtle point: if the user mutated ```message.data``` before calling ```breakLease_```, then ```this.inventory_.bytes``` will now have the wrong value. Do we want to be robust to the user mutating the data? We could retain the value originally added to ```this.inventory_.bytes``` in another field on the message object so that we can guarantee that we subtract the same value here.